connect/runtime/src/test/java/org/apache/kafka/connect/integration/ConnectWorkerIntegrationTest.java

```diff
@@ -926,6 +933,119 @@ public void testPollTimeoutExpiry() throws Exception {
+        );
+        // We verify this task id because it is the one which gets duplicated. Banking upon the assignment logic of
+        // ICR here.
+        String taskIdToVerify = new ConnectorTaskId(CONNECTOR_NAME + "-2", 1).toString();
+
+        // Restarting the task on a separate thread to not block the test thread.
+        Thread restartThread = new Thread(() -> {
+            try {
+                connect.restartTask(CONNECTOR_NAME + "-1", 0);
+            } catch (Exception e) {
+                log.error("Exception while restarting task", e);
+            }
+        });
+        restartThread.start();
+
+        // rebalance.timeout.ms + scheduled.rebalance.delay + 5s buffer.
+        Thread.sleep(Duration.ofSeconds(20).toMillis());
+        List<Map<String, Object>> statuses = new ArrayList<>();
+        try (JsonConverter converter = new JsonConverter()) {
+            ConsumerRecords<byte[], byte[]> records = connect.kafka().consumeAll(Duration.ofSeconds(30).toMillis(), statusTopic);
+            converter.configure(Collections.singletonMap(JsonConverterConfig.SCHEMAS_ENABLE_CONFIG, "false"), false);
+            for (ConsumerRecord<byte[], byte[]> record : records) {
+                String key = new String(record.key());
+                if (!key.equals("status-task-" + taskIdToVerify)) {
+                    continue;
+                }
+                Object valueObject = converter.toConnectData("unused", record.value()).value();
+                @SuppressWarnings("unchecked")
+                Map<String, Object> value = assertAndCast(valueObject, Map.class, "Value");
+                statuses.add(value);
+            }
+        }
+        restartThread.interrupt();
+        String leaderWorkerId = leader.url().getHost() + ":" + leader.url().getPort();
+        String timedOutWorkerId = timingOutWorker.url().getHost() + ":" + timingOutWorker.url().getPort();
+
+        // The task goes through 3 states. RUNNING on first worker, UNASSIGNED on the same worker and then starting on the other
+        // worker. If we had duplicate assignments, because the worker doesn't revoke tasks on poll timeout expiry,
+        // we notice just 2 RUNNING statuses on 2 different workers which means duplicate instances. Note that in some
+        // cases, it could also mean that we couldn't write the UNASSIGNED status to the status topic, but the timeouts
+        // have been set to a high value and I have run the test multiple times to observe the same behaviour.
+        assertEquals(3, statuses.size());
+        assertEquals("RUNNING", statuses.get(0).get("state"));
+        assertEquals(timedOutWorkerId, statuses.get(0).get("worker_id"));
+
+        assertEquals("UNASSIGNED", statuses.get(1).get("state"));
+        assertEquals(timedOutWorkerId, statuses.get(1).get("worker_id"));
+
+        assertEquals("RUNNING", statuses.get(2).get("state"));
+        assertEquals(leaderWorkerId, statuses.get(2).get("worker_id"));
```

Comment:
For reference here is the dump of the status topic with and w/o the fix.
**With revoking assignments on poll timeout**
```
[2024-07-18 11:19:20,217] INFO Record, key::status-connector-simple-connector, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":2} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,217] INFO Record, key::status-task-simple-connector-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":3} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,217] INFO Record, key::status-connector-simple-connector-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50317","generation":5} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-task-simple-connector-1-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50317","generation":6} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-connector-simple-connector-2, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":7} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-task-simple-connector-2-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-task-simple-connector-2-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50317","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-connector-simple-connector-1, value::{"state":"UNASSIGNED","trace":null,"worker_id":"localhost:50317","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,218] INFO Record, key::status-task-simple-connector-2-1, value::{"state":"UNASSIGNED","trace":null,"worker_id":"localhost:50317","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,219] INFO Record, key::status-connector-simple-connector-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,219] INFO Record, key::status-task-simple-connector-1-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:19:20,219] INFO Record, key::status-task-simple-connector-2-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50292","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
```

**Without revoking assignments**
```
[2024-07-18 11:23:07,574] INFO Record, key::status-connector-simple-connector, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":2} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,574] INFO Record, key::status-task-simple-connector-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":3} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,574] INFO Record, key::status-connector-simple-connector-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50409","generation":5} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,574] INFO Record, key::status-task-simple-connector-1-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50409","generation":6} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,574] INFO Record, key::status-connector-simple-connector-2, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":7} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,575] INFO Record, key::status-task-simple-connector-2-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50409","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,575] INFO Record, key::status-task-simple-connector-2-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":8} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,575] INFO Record, key::status-connector-simple-connector-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,575] INFO Record, key::status-task-simple-connector-1-0, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
[2024-07-18 11:23:07,575] INFO Record, key::status-task-simple-connector-2-1, value::{"state":"RUNNING","trace":null,"worker_id":"localhost:50384","generation":10} (org.apache.kafka.connect.integration.ConnectWorkerIntegrationTest:991)
```